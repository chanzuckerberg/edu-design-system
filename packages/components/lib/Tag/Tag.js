"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.stylesByColor = void 0;
const clsx_1 = __importDefault(require("clsx"));
const react_1 = __importDefault(require("react"));
const Text_1 = __importDefault(require("../Text"));
const Tag_module_css_1 = __importDefault(require("./Tag.module.css"));
exports.stylesByColor = {
    neutral: Tag_module_css_1.default.colorNeutral,
    warning: Tag_module_css_1.default.colorWarning,
    brand: Tag_module_css_1.default.colorBrand,
    alert: Tag_module_css_1.default.colorAlert,
    yellow: Tag_module_css_1.default.colorYellow,
    success: Tag_module_css_1.default.colorSuccess,
};
/**
 * This component provides a tag (pill shaped badge) wrapper.
 */
function Tag({ color, children, className, icon, variant = "flat" }) {
    return (react_1.default.createElement(Text_1.default, { as: "span", className: (0, clsx_1.default)(className, Tag_module_css_1.default.tag, color && exports.stylesByColor[color], variant === "outline" && Tag_module_css_1.default.outline), color: color, size: "sm", spacing: "none", weight: "bold" },
        icon,
        "\u200B",
        children && react_1.default.createElement("span", { className: Tag_module_css_1.default.body }, children)));
}
exports.default = Tag;
//# sourceMappingURL=Tag.js.map