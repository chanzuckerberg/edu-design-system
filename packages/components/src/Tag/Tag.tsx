// @flow

import * as React from "react";
import Text from "../Text";
import { TypographyColor } from "../common/typography";
import cx from "classnames";
import styles from "./Tag.module.css";

export type Color = Omit<
  TypographyColor,
  "white" | "base" | "info" | "inherit"
>;

export const stylesByColor: { [K in Color]: string } = {
  neutral: styles.colorNeutral,
  warning: styles.colorWarning,
  brand: styles.colorBrand,
  alert: styles.colorAlert,
  gradingYellow: styles.colorGradingYellow,
  success: styles.colorSuccess,
};

type Props = {
  /**
   * The color of the tag. New colors should be supported in the flow type,
   * in the stylesByColor object, and by its separate style in CSS file.
   */
  color: Color;
  /**
   * The contents of the tag in addition to the icon
   */
  children: React.ReactNode;
  /**
   * The tag icon (shouldn't provide color or size since those are determined
   * by the color prop).
   */
  icon?: React.ReactNode;
  /**
   * Style of the tag.
   */
  variant?: "flat" | "outline";
};

/**
 * This component provides a tag (pill shaped badge) wrapper.
 */
function Tag({ color, children, icon, variant = "flat" }: Props) {
  return (
    <Text
      as="span"
      className={cx(
        styles.tag,
        color && stylesByColor[color],
        variant === "outline" && styles.outline,
      )}
      color={color}
      size="sm"
      spacing="none"
      weight="bold"
    >
      {icon}
      {/* No width space to ensure height of contents */}
      {"\u200B"}
      {children && <span className={styles.body}>{children}</span>}
    </Text>
  );
}

export default Tag;
