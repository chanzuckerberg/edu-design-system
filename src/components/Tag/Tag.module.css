@import '../../design-tokens/mixins.css';

/*------------------------------------*\
    # TAG
\*------------------------------------*/

/**
 * The rounded rectangle container that houses information.
 */
.tag {
  width: max-content;
  height: var(--eds-size-3);
  padding-left: var(--eds-size-1-and-half);
  padding-right: var(--eds-size-1-and-half);

  display: inline-flex;
  align-items: center;

  /* Grabs colors from whichever variant is selected. */
  border: var(--tag-secondary-color) solid var(--eds-theme-border-width);
  /* Rounds the corners of the tag. */
  border-radius: var(--eds-border-radius-full);

  background-color: var(--tag-secondary-color);
  color: var(--tag-primary-color);
  --icon-size-default: 0.875rem;
}

/**
 * If tags are used inline, ensures spacing between tags.
 */
.tag > :not(:last-child) {
  margin-right: var(--eds-size-half);
}

/**
 * Keeps text centered and in one line.
 */
.tag__body {
  flex: 1 1 auto;

  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;

  font: var(--eds-theme-typography-tag);
}

/**
 * Color variants.
 */
:where(.tag--neutral) {
  --tag-primary-color: var(--eds-theme-color-text-neutral-default);
  --tag-secondary-color: var(--eds-theme-color-background-neutral-subtle);
  --tag-outline-color: var(--eds-theme-color-border-neutral-default);
}

:where(.tag--error) {
  --tag-primary-color: var(--eds-theme-color-text-utility-error);
  --tag-secondary-color: var(--eds-theme-color-background-utility-error);
  --tag-outline-color: var(--eds-theme-color-border-utility-error-default);
}

:where(.tag--success) {
  --tag-primary-color: var(--eds-theme-color-text-utility-success);
  --tag-secondary-color: var(--eds-theme-color-background-utility-success);
  --tag-outline-color: var(--eds-theme-color-border-utility-success-default);
}

:where(.tag--warning) {
  --tag-primary-color: var(--eds-theme-color-text-utility-warning);
  --tag-secondary-color: var(--eds-theme-color-background-utility-warning);
  --tag-outline-color: var(--eds-theme-color-border-utility-warning-default);
}

:where(.tag--yield) {
  --tag-primary-color: var(--eds-theme-color-text-grade-revise);
  --tag-secondary-color: var(--eds-theme-color-background-grade-revise-subtle);
  --tag-outline-color: var(--eds-theme-color-border-grade-revise-default);
}

:where(.tag--brand) {
  --tag-primary-color: var(--eds-theme-color-text-brand-default);
  --tag-secondary-color: var(--eds-theme-color-background-brand-primary-default);
  --tag-outline-color: var(--eds-theme-color-border-brand-primary-default);
}

/**
 * Adds outline to the tag.
 */
.tag--outline {
  border-color: var(--tag-outline-color);
}
