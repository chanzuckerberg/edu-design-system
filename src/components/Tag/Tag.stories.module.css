/*------------------------------------*\
    # TAG STORIES
\*------------------------------------*/

.tagList > :not(:last-child) {
  margin-right: var(--eds-size-half);
}
