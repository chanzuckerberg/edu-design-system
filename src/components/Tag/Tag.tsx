import clsx from 'clsx';
import React from 'react';
import styles from './Tag.module.css';
import Text from '../Text';

export const VARIANTS = [
  'neutral',
  'success',
  'yield',
  'error',
  'warning',
  'brand',
] as const;

export type Variant = typeof VARIANTS[number];

type Props = {
  /**
   * The color variant of the tag. New variants should be supported
   * in the VARIANTS array and by its separate style in CSS file.
   */
  variant?: Variant;
  /**
   * CSS class names that can be appended to the component for styling.
   */
  className?: string;
  /**
   * The tag icon (shouldn't provide color or size since those are determined
   * by the color prop).
   */
  icon?: React.ReactNode;
  /**
   * The text contents of the tag, nested inside the component, in addition to the icon.
   */
  text?: React.ReactNode;
  /**
   * Adds an outline for the tag. Defaulted to no outline.
   */
  hasOutline?: boolean;
};

/**
 * ```ts
 * import {Tag} from "@chanzuckerberg/eds";
 * ```
 *
 * This component provides a tag (pill shaped badge) wrapper.
 */
export const Tag = ({
  variant = 'neutral',
  className,
  icon,
  text,
  hasOutline = false,
}: Props) => {
  const componentClassName = clsx(
    className,
    styles['tag'],
    styles[`tag--${variant}`],
    hasOutline && styles['tag--outline'],
  );

  return (
    <Text as="span" className={componentClassName} size="sm" weight="bold">
      {icon}
      {text && <span className={styles['tag__body']}>{text}</span>}
    </Text>
  );
};

export default Tag;
