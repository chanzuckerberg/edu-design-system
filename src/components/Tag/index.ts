export { Tag as default } from './Tag';
