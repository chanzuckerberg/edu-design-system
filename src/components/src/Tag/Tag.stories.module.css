.tagList > :not(:last-child) {
  @apply mr-1;
}
