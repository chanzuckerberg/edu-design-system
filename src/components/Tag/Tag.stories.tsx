import type { StoryObj } from '@storybook/react';
import React from 'react';
import Tag, { VARIANTS } from './Tag';
import styles from './Tag.stories.module.css';
import Icon from '../Icon';

export default {
  title: 'Molecules/Messaging/Tag',
  component: Tag,

  argTypes: {
    variant: {
      control: {
        type: 'select',
        options: VARIANTS,
      },
    },
  },

  args: {
    text: 'Tag text',
    variant: 'default' as const,
  },
};

type Args = React.ComponentProps<typeof Tag>;

export const Default: StoryObj<Args> = {};

export const ColorVariants: StoryObj<Args> = {
  render: (args) => (
    <div className={styles.tagList}>
      {VARIANTS.map((variant) => {
        return <Tag key={variant} {...args} variant={variant} text={variant} />;
      })}
    </div>
  ),
};

export const OutlineVariants: StoryObj<Args> = {
  render: (args) => (
    <div className={styles.tagList}>
      {VARIANTS.map((variant) => {
        return (
          <Tag
            key={variant}
            {...args}
            variant={variant}
            text={variant}
            hasOutline={true}
          />
        );
      })}
    </div>
  ),
};

/**
 * Snap tests disabled due to TypeError in <Icon> but should be turned back on once fixed.
 */
export const WithIcon: StoryObj<Args> = {
  ...Default,
  args: {
    icon: <Icon purpose="decorative" name="favorite" />,
  },
  parameters: {
    snapshot: { skip: true },
  },
  render: (args) => (
    <div className={styles.tagList}>
      {VARIANTS.map((variant) => {
        return (
          <Tag
            key={variant}
            {...args}
            variant={variant}
            text={variant}
            hasOutline={true}
          />
        );
      })}
    </div>
  ),
};

export const WithLongTextAndIcon: StoryObj<Args> = {
  ...Default,
  args: {
    text: 'This tag has a really long text message',
    icon: <Icon purpose="decorative" name="star" />,
  },
  parameters: {
    snapshot: { skip: true },
  },
  render: (args) => (
    <div className={styles.tagList}>
      {VARIANTS.map((variant) => {
        return (
          <Tag key={variant} {...args} variant={variant} hasOutline={true} />
        );
      })}
    </div>
  ),
};
