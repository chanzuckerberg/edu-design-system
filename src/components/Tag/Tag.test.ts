import { generateSnapshots } from '@chanzuckerberg/story-utils';
import * as stories from './Tag.stories';

describe('<Tag />', () => {
  generateSnapshots(stories);
});
