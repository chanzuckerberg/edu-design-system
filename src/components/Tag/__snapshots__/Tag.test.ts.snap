// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tag /> ColorVariants story renders snapshot 1`] = `
<div
  class="tagList"
>
  <span
    class="text text--sm text--bold-weight tag tag--neutral"
    data-testid="test"
  >
    <span
      class="tag__body"
    >
      neutral
    </span>
  </span>
  <span
    class="text text--sm text--bold-weight tag tag--error"
    data-testid="test"
  >
    <span
      class="tag__body"
    >
      error
    </span>
  </span>
  <span
    class="text text--sm text--bold-weight tag tag--success"
    data-testid="test"
  >
    <span
      class="tag__body"
    >
      success
    </span>
  </span>
  <span
    class="text text--sm text--bold-weight tag tag--warning"
    data-testid="test"
  >
    <span
      class="tag__body"
    >
      warning
    </span>
  </span>
  <span
    class="text text--sm text--bold-weight tag tag--yield"
    data-testid="test"
  >
    <span
      class="tag__body"
    >
      yield
    </span>
  </span>
  <span
    class="text text--sm text--bold-weight tag tag--brand"
    data-testid="test"
  >
    <span
      class="tag__body"
    >
      brand
    </span>
  </span>
</div>
`;

exports[`<Tag /> Default story renders snapshot 1`] = `
<span
  class="text text--sm text--bold-weight tag tag--neutral"
>
  <span
    class="tag__body"
  >
    Tag text
  </span>
</span>
`;

exports[`<Tag /> OutlineVariants story renders snapshot 1`] = `
<div
  class="tagList"
>
  <span
    class="text text--sm text--bold-weight tag tag--neutral tag--outline"
  >
    <span
      class="tag__body"
    >
      neutral
    </span>
  </span>
  <span
    class="text text--sm text--bold-weight tag tag--error tag--outline"
  >
    <span
      class="tag__body"
    >
      error
    </span>
  </span>
  <span
    class="text text--sm text--bold-weight tag tag--success tag--outline"
  >
    <span
      class="tag__body"
    >
      success
    </span>
  </span>
  <span
    class="text text--sm text--bold-weight tag tag--warning tag--outline"
  >
    <span
      class="tag__body"
    >
      warning
    </span>
  </span>
  <span
    class="text text--sm text--bold-weight tag tag--yield tag--outline"
  >
    <span
      class="tag__body"
    >
      yield
    </span>
  </span>
  <span
    class="text text--sm text--bold-weight tag tag--brand tag--outline"
  >
    <span
      class="tag__body"
    >
      brand
    </span>
  </span>
</div>
`;

exports[`<Tag /> WithIcon story renders snapshot 1`] = `
<div
  class="tagList"
>
  <span
    class="text text--sm text--bold-weight tag tag--neutral tag--outline"
  >
    <svg
      aria-hidden="true"
      class="icon"
      fill="currentColor"
      xmlns="http://www.w3.org/2000/svg"
    >
      <use
        xlink:href="test-file-stub#favorite"
      />
    </svg>
    <span
      class="tag__body"
    >
      neutral
    </span>
  </span>
  <span
    class="text text--sm text--bold-weight tag tag--error tag--outline"
  >
    <svg
      aria-hidden="true"
      class="icon"
      fill="currentColor"
      xmlns="http://www.w3.org/2000/svg"
    >
      <use
        xlink:href="test-file-stub#favorite"
      />
    </svg>
    <span
      class="tag__body"
    >
      error
    </span>
  </span>
  <span
    class="text text--sm text--bold-weight tag tag--success tag--outline"
  >
    <svg
      aria-hidden="true"
      class="icon"
      fill="currentColor"
      xmlns="http://www.w3.org/2000/svg"
    >
      <use
        xlink:href="test-file-stub#favorite"
      />
    </svg>
    <span
      class="tag__body"
    >
      success
    </span>
  </span>
  <span
    class="text text--sm text--bold-weight tag tag--warning tag--outline"
  >
    <svg
      aria-hidden="true"
      class="icon"
      fill="currentColor"
      xmlns="http://www.w3.org/2000/svg"
    >
      <use
        xlink:href="test-file-stub#favorite"
      />
    </svg>
    <span
      class="tag__body"
    >
      warning
    </span>
  </span>
  <span
    class="text text--sm text--bold-weight tag tag--yield tag--outline"
  >
    <svg
      aria-hidden="true"
      class="icon"
      fill="currentColor"
      xmlns="http://www.w3.org/2000/svg"
    >
      <use
        xlink:href="test-file-stub#favorite"
      />
    </svg>
    <span
      class="tag__body"
    >
      yield
    </span>
  </span>
  <span
    class="text text--sm text--bold-weight tag tag--brand tag--outline"
  >
    <svg
      aria-hidden="true"
      class="icon"
      fill="currentColor"
      xmlns="http://www.w3.org/2000/svg"
    >
      <use
        xlink:href="test-file-stub#favorite"
      />
    </svg>
    <span
      class="tag__body"
    >
      brand
    </span>
  </span>
</div>
`;

exports[`<Tag /> WithLongTextAndIcon story renders snapshot 1`] = `
<div
  class="tagList"
>
  <span
    class="text text--sm text--bold-weight tag tag--neutral tag--outline"
  >
    <svg
      aria-hidden="true"
      class="icon"
      fill="currentColor"
      xmlns="http://www.w3.org/2000/svg"
    >
      <use
        xlink:href="test-file-stub#star"
      />
    </svg>
    <span
      class="tag__body"
    >
      This tag has a really long text message
    </span>
  </span>
  <span
    class="text text--sm text--bold-weight tag tag--error tag--outline"
  >
    <svg
      aria-hidden="true"
      class="icon"
      fill="currentColor"
      xmlns="http://www.w3.org/2000/svg"
    >
      <use
        xlink:href="test-file-stub#star"
      />
    </svg>
    <span
      class="tag__body"
    >
      This tag has a really long text message
    </span>
  </span>
  <span
    class="text text--sm text--bold-weight tag tag--success tag--outline"
  >
    <svg
      aria-hidden="true"
      class="icon"
      fill="currentColor"
      xmlns="http://www.w3.org/2000/svg"
    >
      <use
        xlink:href="test-file-stub#star"
      />
    </svg>
    <span
      class="tag__body"
    >
      This tag has a really long text message
    </span>
  </span>
  <span
    class="text text--sm text--bold-weight tag tag--warning tag--outline"
  >
    <svg
      aria-hidden="true"
      class="icon"
      fill="currentColor"
      xmlns="http://www.w3.org/2000/svg"
    >
      <use
        xlink:href="test-file-stub#star"
      />
    </svg>
    <span
      class="tag__body"
    >
      This tag has a really long text message
    </span>
  </span>
  <span
    class="text text--sm text--bold-weight tag tag--yield tag--outline"
  >
    <svg
      aria-hidden="true"
      class="icon"
      fill="currentColor"
      xmlns="http://www.w3.org/2000/svg"
    >
      <use
        xlink:href="test-file-stub#star"
      />
    </svg>
    <span
      class="tag__body"
    >
      This tag has a really long text message
    </span>
  </span>
  <span
    class="text text--sm text--bold-weight tag tag--brand tag--outline"
  >
    <svg
      aria-hidden="true"
      class="icon"
      fill="currentColor"
      xmlns="http://www.w3.org/2000/svg"
    >
      <use
        xlink:href="test-file-stub#star"
      />
    </svg>
    <span
      class="tag__body"
    >
      This tag has a really long text message
    </span>
  </span>
</div>
`;
